docker/elasticsearch-setup/create-indices.sh

```diff
@@ -14,22 +14,110 @@ else
+    --slurpfile mappings "$2" \
+    '.settings=$settings[0] | .mappings=$mappings[0]' > "$3"
+}
+
+function check_reindex() {
+    initial_documents=$(curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$1/_count" -H 'Content-Type: application/json' | jq '.count')
+    for i in {1..30}; do
+      reindexed_documents=$(curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$2/_count" -H 'Content-Type: application/json' | jq '.count')
+      if [[ $reindexed_documents == "$initial_documents" ]]; then
+        echo -e "\nPost-reindex document reconcialiation completed. doc_source_index_count: $initial_documents; doc_target_index_count: $reindexed_documents"
+        return 0
+      else
+        sleep 3
+      fi
+    done
+
+    echo -e "\nPost-reindex document reconcialiation failed. doc_source_index_count: $initial_documents; doc_target_index_count: $reindexed_documents"
+    return 1
+}
+
+function reindex() {
+  source_index=$1
+  target_index="$1_$(date +%s)"
+
+  #create target index with latest index config
+  curl -XPUT "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$target_index" -H 'Content-Type: application/json' --data @/tmp/data
+
+  #reindex the documents in source index to target index.
+  # One of the assumption here is that we only add properties to document when index-config is evolved.
+  # In case a property is deleted from document, it will still be reindexed in target index as default behaviour and
+  # it is not breaking the code. If still needs to be purged from target index, use "removed" property in POST data.
+  curl -XPOST "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_reindex?pretty" -H 'Content-Type: application/json' \
+    -d "{\"source\":{\"index\":\"$source_index\"},\"dest\":{\"index\":\"$target_index\"}}"
+
+  if check_reindex "$source_index" "$target_index"
+  then
+    #checking if source index is concrete index or alias
+    if [ $(curl -o /dev/null -s -w "%{http_code}" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_alias/$source_index") -eq 404 ]
+    then
+      curl -XDELETE "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$source_index"
+    else
+      concrete_index_name=$(curl -XGET "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_alias/$source_index" | jq 'keys[]' | head -1 | tr -d \")
+      curl -XDELETE "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/$concrete_index_name"
+    fi
+
+    curl -XPOST "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST_URL:$ELASTICSEARCH_PORT/_aliases" -H 'Content-Type: application/json' \
+      -d "{\"actions\":[{\"remove\":{\"index\":\"*\",\"alias\":\"$source_index\"}},{\"add\":{\"index\":\"$target_index\",\"alias\":\"$source_index\"}}]}"
+
+    echo -e "\nReindexing to $target_index succeded"
+    return 0
+  else
+    echo -e "\nReindexing to $target_index failed"
```

Comment:
We should delete target_index on failure.

Reply:
Done. Thanks for feedback.